src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+  public validate(value: AcceptedNumber): true {
+    if (!Validator.isDecimal(String(value))) {
+      throw new ValidationError(
+        util.format('%j is not a valid decimal', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * Decimal type, variable precision, take length as specified by user
+ */
+export type DECIMAL = _DECIMAL;
+export const DECIMAL = classToInvokable(_DECIMAL);
+
+/**
+ * A boolean / tinyint column, depending on dialect
+ */
+class _BOOLEAN extends ABSTRACT<
+  boolean | Falsy,
+  boolean,
+  string | number | Buffer | boolean
+> {
+  public static readonly key = 'BOOLEAN';
+
+  public toSql() {
+    // Note: This may vary depending on the dialect.
+    return 'TINYINT(1)';
+  }
+
+  public validate(value: boolean | Falsy): asserts value is boolean;
+  public validate(value: boolean | Falsy): true {
+    if (!Validator.isBoolean(String(value))) {
+      throw new ValidationError(
+        util.format('%j is not a valid boolean', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+
+  protected _sanitize(
+    value: string | number | Buffer | boolean,
+  ): boolean | null {
+    if (value !== null && value !== undefined) {
+      if (Buffer.isBuffer(value) && value.length === 1) {
+        // Bit fields are returned as buffers
+        value = value[0];
+      }
+
+      const type = typeof value;
+      if (type === 'string') {
+        // Only take action on valid boolean strings.
+        if (value === 'true') {
+          return true;
+        }
+
+        if (value === 'false') {
+          return false;
+        }
+      } else if (
+        type === 'number' // Only take action on valid boolean integers.
+        && (value === 0 || value === 1)
+      ) {
+        return Boolean(value);
+      }
+    }
+
+    return null;
+  }
+
+  static readonly parse = this.prototype._sanitize;
+}
+
+export type BOOLEAN = _BOOLEAN;
+export const BOOLEAN = classToInvokable(_BOOLEAN);
+
+class _TIME extends ABSTRACT<Date | string | number, Date> {
+  public static readonly key = 'TIME';
+
+  toSql() {
+    return 'TIME';
+  }
+}
+
+/**
+ * A time column
+ */
+export type TIME = _TIME;
+export const TIME = classToInvokable(_TIME);
+
+interface DateOptions {
+  /**
+   * The precision of the date.
+   */
+  length?: string | number;
+}
+
+type RawDate = Date | string | number;
+type AcceptedDate = RawDate | moment.Moment;
+
+class _DATE extends ABSTRACT<AcceptedDate, Date, RawDate> {
+  public static readonly key: string = 'DATE';
+  protected options: DateOptions;
+  protected _length?: string | number;
+
+  /**
+   * @param [length] precision to allow storing milliseconds
+   */
+  constructor(length?: number | DateOptions) {
+    super();
+    const options: DateOptions
+      = typeof length === 'object' ? length : { length };
+    this.options = options;
+    // TODO: Check if this is still used and remove if it isn't.
+    this._length = options.length || '';
```

Comment:
same comment about deduplicating `options` & instance fields

'_length' is probably used for the precision parameter most dialects support.